addons/product/static/src/js/product_pricelist_report.js

```diff
@@ -1,317 +1,216 @@
+import { Layout } from "@web/search/layout";
+
+const { Component, useState, onWillStart, markup } = owl;
+
+class GeneratePriceListReport extends Component {
+
+    setup() {
+        this.orm = useService("orm");
         this.MAX_QTY = 5;
-    },
-
-    //--------------------------------------------------------------------------
-    // Handlers
-    //--------------------------------------------------------------------------
-
-    /**
-     * Add a quantity when add(+) button clicked.
-     *
-     * @private
-     */
-    _onClickAddQty: function () {
-        if (this.quantities.length >= this.MAX_QTY) {
-            this.displayNotification({ message: _.str.sprintf(
-                _t("At most %d quantities can be displayed simultaneously. Remove a selected quantity to add others."),
-                this.MAX_QTY
-            ) });
+        this.action = useService("action");
+        this.active_model = this.props.action.context.active_model;
+        this.active_ids = this.props.action.context.active_ids;
+
+        const pastState = this.props.state || {};
+
+        this.state = useState({
+            quantities: pastState.quantities || [1, 5, 10],
+            displayPricelist: pastState.displayPricelist || false,
+            pricelists: [],
+            selectedPricelist: pastState.selectedPricelist || {},
+            html: markup(""),
+        });
+
+        onWillStart(async () => {
+            this.state.pricelists = await this.getPricelist()
+            this.state.selectedPricelist = this.state.selectedPricelist || this.state.pricelists[0];
+
+            this.renderHtml();
+       });
+
+       useSetupAction({
+            getLocalState: () => {
+                return {
+                    displayPricelist: this.state.displayPricelist,
+                    selectedPricelist: this.state.selectedPricelist,
+                    quantities: this.state.quantities,
+                };
+            },
+        });
+    }
+
+    // getters
+
+   get quantities() {
+        return this.state.quantities;
+    }
+
+    get html() {
+        return this.state.html;
+    }
+
+    get displayPricelist() {
+        return this.state.displayPricelist;
+    }
+
+    get pricelists() {
+        return this.state.pricelists;
+    }
+
+    get selectedPricelist() {
+        return this.state.selectedPricelist;
+    }
+
+    get noSelectedPricelists() {
+        return this.state.pricelists.filter(
+            pricelist => pricelist.id !== this.state.selectedPricelist.id
+        )
+    }
+
+    get getReportParams() {
+        return {
+            active_model: this.active_model || 'product.template',
+            active_ids: this.active_ids || '',
+            is_visible_title: this.displayPricelist || '',
+            pricelist_id: this.state.selectedPricelist.id || '',
+            quantities: this.state.quantities || [1],
+        };
+    }
+
+    // orm calls
+
+    async renderHtml() {
+        let html = await this.getHtml();
```

Comment:
As this methode is only called here, do you really need to have a specific `getHtml` method? Can't you make the RPC the code here.?

Reply:
How so? `getHtml` is called only once. The advantage of keeping it all in the same method is that it's easier to detect if one touches the content of `html` to potentially introduce an XSS.